packages/core/src/components/forms/inputGroup.tsx

```diff
@@ -133,11 +143,24 @@ export class InputGroup extends AbstractPureComponent2<IInputGroupProps & HTMLIn
     }
 
     public componentDidUpdate(prevProps: IInputGroupProps & HTMLInputProps) {
-        if (prevProps.rightElement !== this.props.rightElement) {
+        const { leftElement, rightElement } = this.props;
+        if (prevProps.leftElement !== leftElement || prevProps.rightElement !== rightElement) {
             this.updateInputWidth();
         }
     }
 
+    private maybeRenderLeftElement() {
+        const { leftElement, leftIcon } = this.props;
+        if (leftElement == null) {
+            return <Icon icon={leftIcon} />;
+        }
```

Comment:
I think we should return `undefined` if there is no leftElement. more concretely:

```
if (leftElement != null) {
  ...
} else if (leftIcon != null) {
  ...
}
```

this also matches our naming convention where "maybeX" render methods always have `| undefined` in their computed return type.

in addition, I think you should use prop validation to ensure `leftElement` and `leftIcon` are mutually exclusive. we do this with `AbstractPureComponent2#validateProps`. you will need to implement `validateProps` in this component. also make a note of this in the props JSDoc comments

Reply:
Done, thank you for the input